Clarify usage of ActivityPub\Transmitter::fetchTargetInboxes in Worker\APDelivery queuing

MrPetovan opened the pull request:
Follow-up to #9651 
Address https://github.com/friendica/friendica/issues/9250#issuecomment-744737015

The original PR changed `$inboxes` returned by `ActivityPub\Transmitter::fetchTargetInboxes` from a list of contact URLs to an associative array of contact ID lists indexed by their URL. Its use has been updated in a single place, but not everywhere it was used, leading to data discrepancies caught by PHP type checking.

This PR reverts part of https://github.com/friendica/friendica/pull/9655/commits/481c5c9e81f0eded399c5bdc3c25b27f7a628018 that incorrectly assumed that the `$inbox` parameter of the worker task `APDelivery` was supposed to be an array when it was still supposed to be a string.

The problem is that now several admins have worker tasks with corrupted data in worker task parameters that will systematically trip PHP type-check and whose parameters must be manually updated before they can be correctly processed. However, since it concerns only admins running the `2020.12-rc`, any data correction script must not make it into the upcoming release.

MrPetovan committed:
Clarify usage of ActivityPub\Transmitter::fetchTargetInboxes in Worker\APDelivery queuing

- $inbox is now always a string
- $receivers should be correctly populated now
- $target_id is renamed to $item_id and correctly initialized to 0 for non-item deliveries

MrPetovan committed:
Add temporary console command to fix workerqueue row with corrupted parameter data

- It should be reverted just before the 2020.12 release

MrPetovan commented:
I added a console command to fix the parameters of the corrupted workerqueue entries that should be removed before release. Since my production node hasn't been affected, I can't guarantee it will correct all the affected rows. I've been working with [the limited dataset](https://github.com/friendica/friendica/issues/9250#issuecomment-744836857) @AlfredSK kindly provided from his node.

annando commented:
My current time is slightly limited. I will try having a look at it this evening. Are you working with this branch on your productive machine?

MrPetovan commented:
No, like I said I never updated to the faulty commit which means I never got any corrupted data. But I stand by the other changes (particularly the foreach changes after `fetchInboxes`).

AlfredSK commented:
I could try it. Is it possible to switch my upstream to MrPetovan's repo and then switching the branch to `9651-apdelivery-inboxes`? Will this work?

AlfredSK commented:
Okay. Switched upstream repo and branch on venera.social. Let's see...

```
sudo -u www-data bin/console fixapdeliveryworkertaskparameters -v
Executable: /var/www/html/venera/bin/console.php
Arguments: array (
  0 => 'fixapdeliveryworkertaskparameters',
)
Options: array (
  'v' => 1,
)
Command: fixapdeliveryworkertaskparameters
Class: Friendica\Console\FixAPDeliveryWorkerTaskParameters
Arguments: array (
)
Options: array (
  'v' => 1,
)
Examined: 2300
Processed: 2300
Errored: 0
```

Now I will watch the workerqueue....

AlfredSK commented:
I'm still getting the two fatal errors:

```
PHP Fatal error:  Uncaught TypeError: Argument 2 passed to Friendica\Worker\APDelivery::execute() must be of the type int, string given in /var/www/html/venera/src/Worker/APDelivery.php:45
Stack trace:
#0 [internal function]: Friendica\Worker\APDelivery::execute()
#1 /var/www/html/venera/src/Core/Worker.php(438): call_user_func_array()
#2 /var/www/html/venera/src/Core/Worker.php(334): Friendica\Core\Worker::execFunction()
#3 /var/www/html/venera/src/Core/Worker.php(105): Friendica\Core\Worker::execute()
#4 /var/www/html/venera/bin/worker.php(81): Friendica\Core\Worker::processQueue()
#5 {main}
  thrown in /var/www/html/venera/src/Worker/APDelivery.php on line 45
```

and

```
PHP Fatal error:  Uncaught TypeError: Argument 5 passed to Friendica\Worker\APDelivery::execute() must be of the type array, string given in /var/www/html/venera/src/Worker/APDelivery.php:45
Stack trace:
#0 [internal function]: Friendica\Worker\APDelivery::execute()
#1 /var/www/html/venera/src/Core/Worker.php(438): call_user_func_array()
#2 /var/www/html/venera/src/Core/Worker.php(334): Friendica\Core\Worker::execFunction()
#3 /var/www/html/venera/src/Core/Worker.php(105): Friendica\Core\Worker::execute()
#4 /var/www/html/venera/bin/worker.php(81): Friendica\Core\Worker::processQueue()
#5 {main}
  thrown in /var/www/html/venera/src/Worker/APDelivery.php on line 45
```

AlfredSK commented:
I had to run the console command several times. The numbers went down with each round. Now I get...

```
sudo -u www-data bin/console fixapdeliveryworkertaskparameters -v
Executable: /var/www/html/venera/bin/console.php
Arguments: array (
  0 => 'fixapdeliveryworkertaskparameters',
)
Options: array (
  'v' => 1,
)
Command: fixapdeliveryworkertaskparameters
Class: Friendica\Console\FixAPDeliveryWorkerTaskParameters
Arguments: array (
)
Options: array (
  'v' => 1,
)
Examined: 0
Processed: 0
Errored: 0
```

MrPetovan commented:
> I had to run the console command several times. 

It's normal, your node is still adding new corrupted tasks because of a running daemon or background process with the faulty code. As the daemon will be restarted/background workers will be replaced, the new code will be used.

AlfredSK commented:
The queued 'remove me' items have been processed. They're now in the 'Deferred Worker Queue'. I guess that's the expected behavior.

MrPetovan commented:
Phew!

AlfredSK commented:
I will let it run on the branch for some hours and see if the fatal errors still occur. I also switched libranet.de to that branch now. :-)

AlfredSK commented:
Until now I only see some delivery issues for 'wall-new' items sent to the AP relays. I need to investigate this.

To have a reliable baseline I removed the AP relays and deleted all workerqueue items for the relays. Also scrubbed everything from the relay itself. Then I added one AP relay again. That worked. Now I can see how the delivery works.

annando commented:
I'm currently confused about what github shows as the previous code and your changes - because the old code doesn't reflect the current state. Could you make it more clear what you really changed? Because now it looks like your changes would be the inbox arrays - but this is already in the system.

MrPetovan commented:
The crux of the issue is that the APDelivery task shouldn't accept an array for the inbox parameter. It's just one task per inbox. What changed is actually outside of the `Worker\APDelivery` class. I've fixed how we loop through the new inbox array to queue worker task. So the real changes are in `Worker\Notifier` and `Worker\ProfileUpdate`.

The rest is reverting my earlier changes to `APDelivery` and the temporary console command.

AlfredSK commented:
Question: Does this look right now? I'm confused about the combination of 'executed', 'done' and 'next_try'. Thing is, these workerqueue entries seem to stay in the queue forever.

```
+-----------+-----------------------------------------------------------------------------------------------------------+----------+---------------------+--------+---------------------+------+---------------------+---------+------------+
| id        | parameter                                                                                                 | priority | created             | pid    | executed            | done | next_try            | retrial | command    |
+-----------+-----------------------------------------------------------------------------------------------------------+----------+---------------------+--------+---------------------+------+---------------------+---------+------------+
| 218346337 | ["wall-new",101523015,"https:\/\/relay.libranet.de\/inbox",3822,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:03:34 | 174131 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218346418 | ["wall-new",101522806,"https:\/\/relay.libranet.de\/inbox",3822,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:03:35 | 174137 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218346458 | ["wall-new",101523112,"https:\/\/relay.libranet.de\/inbox",3868,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:04:02 | 174139 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218346921 | ["wall-new",101523233,"https:\/\/relay.libranet.de\/inbox",3822,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:04:37 | 174141 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218347076 | ["wall-new",101523282,"https:\/\/relay.libranet.de\/inbox",3822,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:05:08 | 174141 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218347129 | ["wall-new",101523317,"https:\/\/relay.libranet.de\/inbox",3822,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:05:17 | 174141 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218347151 | ["wall-new",101523354,"https:\/\/relay.libranet.de\/inbox",3258,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:05:40 | 174143 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218347365 | ["wall-new",101523616,"https:\/\/relay.libranet.de\/inbox",3822,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:06:55 | 174143 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218347541 | ["wall-new",101524134,"https:\/\/relay.libranet.de\/inbox",3701,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:09:40 | 174150 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218347924 | ["wall-new",101524819,"https:\/\/relay.libranet.de\/inbox",3258,"https:\/\/relay.libranet.de\/inbox"]     |       20 | 2020-12-15 20:15:40 | 174150 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
| 218347925 | ["wall-new",101524819,"https:\/\/de.relay.friendi.ca\/inbox",3258,"https:\/\/de.relay.friendi.ca\/inbox"] |       20 | 2020-12-15 20:15:40 | 174150 | 2020-12-15 20:21:00 |    0 | 0001-01-01 00:00:00 |       0 | APDelivery |
+-----------+-----------------------------------------------------------------------------------------------------------+----------+---------------------+--------+---------------------+------+---------------------+---------+------------+
```

MrPetovan commented:
No, the 5th parameter shouldn't be the inbox URL again. It should be an array of integers like `[104534,2345257,262345]`. You should be getting type check fatal errors for these which would indeed prevent their processing forever.

AlfredSK commented:
To get the type checks I need to run the worker manually again. Wait... :-)

MrPetovan commented:
Never mind, I found ~my~ @annando's error 😜.

AlfredSK commented:
Yup. There they are...

```
PHP Fatal error:  Uncaught TypeError: Argument 5 passed to Friendica\Worker\APDelivery::execute() must be of the type array, string given in /var/www/html/venera/src/Worker/APDelivery.php:45
Stack trace:
#0 [internal function]: Friendica\Worker\APDelivery::execute()
#1 /var/www/html/venera/src/Core/Worker.php(438): call_user_func_array()
#2 /var/www/html/venera/src/Core/Worker.php(334): Friendica\Core\Worker::execFunction()
#3 /var/www/html/venera/src/Core/Worker.php(105): Friendica\Core\Worker::execute()
#4 /var/www/html/venera/bin/worker.php(81): Friendica\Core\Worker::processQueue()
#5 {main}
  thrown in /var/www/html/venera/src/Worker/APDelivery.php on line 45
```

MrPetovan committed:
Remove reference to receivers when adding relay delivery tasks in Worker\Notifier

- Address https://github.com/friendica/friendica/pull/9661#issuecomment-745553051

MrPetovan commented:
If you don't have too many of these, once you applied the patch I just committed, please edit these task manually to remove the duplicate inbox URL in the parameter field.

KO: `["wall-new",101523015,"https:\/\/relay.libranet.de\/inbox",3822,"https:\/\/relay.libranet.de\/inbox"]`
OK: `["wall-new",101523015,"https:\/\/relay.libranet.de\/inbox",3822]`

AlfredSK commented:
My SQL-Fu is limited. I tried it with the 1st row and it was difficult. But I was successful. Now I can see what happens.

oops... I need to double escape the slashes. :-)

annando commented:
Is it currently in a mergeable state?

AlfredSK commented:
Now the relay items are processed and are `done = 1`. :-) Looks good.

AlfredSK commented:
From my point of view it is ready now. Worker is not throwing any error now.

utzer commented:
Can someone help me with the dropping of these items in the queue? I have CLI access and can run some query or so. I still got many of these dead jobs in the queue and would love to get rid of them. Shouldn't old jobs/items in the queue be discarded after some time?

MrPetovan commented:
How many of these rows do you have?

utzer commented:
> How many of these rows do you have?

Maybe 300 on one server and 130 on the other server.

AlfredSK commented:
From my point of view you can just delete them from the queue.

`delete from workerqueue where parameter like '%relay%relay%';`

Maybe you should try this approach with a `select` query first because my SQL knowledge is limited.

utzer commented:
Mine is worse. 
```ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'from workerqueue where parameter like '%relay%relay%'' at line 1```

I replaced delete with select.

AlfredSK commented:
But select doesn't delete. ;-) I cannot see a syntax error. You can try to enclose the table name in backticks. 
Have you select the DB with `use dbname` before executing the query?

I'm pretty sure the syntax is correct because it is working here.

MrPetovan commented:
@ben-utzer If you want to test a `DELETE` query before performing it, you have to replace it with `SELECT *`. The star indicates which columns you want in your result, in this case all of them.

utzer commented:
> Have you select the DB with `use dbname` before executing the query?

I called `mysql dbname`, so that shouldn't be needed, but it is a fault I could have done. (probably did before)

> `DELETE` query before performing it, you have to replace it with `SELECT *`. The star indicates which columns you want in your result, in this case all of them.

That looks good, I get `138 rows in set (0.008 sec)`. Thanks so much to both of you!

utzer commented:
Just another question, in the "defered" part of the queue I have also many jobs from 12th December, so about 14 days old. Is this normal? Like this: 
```
77468577 | drop: 12318165: {15878: 15886: 15929: 16129: 16186: 17256: 17572:  17597: 18895: 19055: 19074: 19181: 20166: 20412: 21430: 21436: 21495:  22211: 22514: 22909: 23149: 23187: 23234: 124748: 130907: 132692:  133738: 134261: 134851: 135740: 135825: 135844: 136675: 137473: 137681:  139856: 139904: 150165: 152413: 153657: 154260: 154732: 154741: 154847:  156003: 156317: 156558: 156688: 156903: 157598: 157600: 157936: 158720:  159502: 159584: 160718: 162263: 163374: 170912: 175981: 179309: 180551:  188077: 188168: 188761: 188845: 189701: 189854: 191293: 191439: 191992:  192318: 192848: 192906: 197205: 197262: 198716: 234371: 234695: 235035:  235560: 239835: 243537: 247960: 537760: 541731: 548568}: 11 | 2020-12-13 02:06:12 | 50
-- | -- | -- | --
77468578 | drop: 12318165: {16024: 16164: 16165: 16207: 16212: 16279: 16298:  16300: 16454: 16569: 17906: 19004: 131249: 132234: 156284: 185548:  236939}: 11 | 2020-12-13 02:06:12 | 50
77468580 | drop: 12318165: {16183: 16409: 16471: 191251: 247143}: 11 | 2020-12-13 02:06:12 | 50
77468581 | drop: 12318165: {16474: 17317: 22297: 129992: 132042: 167954: 170977: 176605: 176634: 191555}: 11 | 2020-12-13 02:06:12 | 50
```

AlfredSK commented:
You shouldn't delete too much from the workerqueue. The malformed items, yes. But the deferred queue doesn't do any harm. The remote nodes couldn't be reached to process these queued jobs. We are trying to process them in increasing intervals. Then - if still not successful - they're discarded and dropped.

utzer commented:
> Then - if still not successful - they're discarded and dropped.

Well I know, so it is normal to keep these jobs for 14 days, then I am fine. 
I just thought that these were relay jobs too.
